kie-remote/kie-remote-client/src/test/java/org/kie/remote/client/ws/AuthenticatorTest.java

```diff
@@ -0,0 +1,352 @@
+ */
+public class AuthenticatorTest {
+
+    private static final Logger logger = LoggerFactory.getLogger(AuthenticatorTest.class);
+
+    private static final String DEFAULT_ENDPOINT = "/ws/CommandService?wsdl";
+    private static final String DEFAULT_REAL_ENDPOINT = "/ws/CommandService";
+
+    private static int port = -1;
+    private TestServer server = null;
+
+    @BeforeClass
+    public static void setup() {
+        port = AvailablePortFinder.getNextAvailable(1025);
+    }
+
+    @After
+    public void cleanup() throws Exception {
+        server.stop();
+    }
+
+    private static interface TestServer {
+        public void start() throws Exception;
+
+        public void stop() throws Exception;
+    }
+
+    public static String readInputStreamAsString( InputStream in ) throws IOException {
+        BufferedInputStream bis = new BufferedInputStream(in);
+        ByteArrayOutputStream buf = new ByteArrayOutputStream();
+        int result = bis.read();
+        while( result != -1 ) {
+            byte b = (byte) result;
+            buf.write(b);
+            result = bis.read();
+        }
+        return buf.toString();
+    }
+
+    private static final String BAD_USER = "bad";
+    private static final String BAD_PASSWORD  = "easy";
+    private static final String GOOD_USER = "good-guy";
+    private static final String GOOD_PASSWORD  = "a4os";
+
+    /**
+     * Scenario to test:
+     * 1. Bad client creation : fails
+     * 2. good client creation: succeeds
+     * 3. good client request : succeeds
+     * 4. bad client creation : SUCCEEDS BUT SHOULD FAIL!
+     *
+     * @throws Exception
+     */
+    @Test
+    public void statefulAuthorizationTest() throws Exception {
+        server = new BasicAuthServer(port);
+        server.start();
+
+        // bad client fails
+        boolean exceptionThrown = false;
+        try {
+            createClient(BAD_USER, BAD_PASSWORD);
+        } catch( RemoteCommunicationException rce) {
+            exceptionThrown = true;
+            assertTrue( rce.getMessage().contains("Status 401"));
+        }
+        assertTrue( "Exception should have been thrown upon client creation with bad login info", exceptionThrown );
+
+        // good client creation
+        CommandWebService cmdServiceClient = createClient(GOOD_USER, GOOD_PASSWORD);
+
+        // good client request
+        JaxbCommandResponse cmdResp = cmdServiceClient.execute(new JaxbCommandsRequest()).getResponses().get(0);
+
+        String user = ((JaxbStringListResponse) cmdResp).getResult().get(0);
+        String password = ((JaxbStringListResponse) cmdResp).getResult().get(1);
+        assertEquals("User did not match request!", GOOD_USER, user);
+        assertEquals("Password did not match request!", GOOD_PASSWORD, password);
+
+        // bad client creation
+        exceptionThrown = false;
+        try {
+            createClient(BAD_USER, BAD_PASSWORD);
```

Comment:
let's call a command here via execute method and that should make the authentication fail with the fix and succeed without the fix